framework/tlvf/src/include/tlvf/WSC/AttrList.h

```diff
@@ -0,0 +1,139 @@
+    sMacAddr mac;
+    uint8_t enrollee_nonce[WSC_NONCE_LENGTH];
+    uint8_t registrar_nonce[WSC_NONCE_LENGTH];
+    uint8_t pub_key[WSC_PUBLIC_KEY_LENGTH];
+    uint16_t encr_type_flags;
+    uint16_t auth_type_flags;
+    std::string manufacturer;
+    std::string model_name;
+    std::string model_number;
+    std::string serial_number;
+    uint16_t primary_dev_type_id;
+    std::string device_name;
+    eWscRfBands bands;
+    std::vector<uint8_t> encrypted_settings;
+    uint8_t iv[WSC_ENCRYPTED_SETTINGS_IV_LENGTH];
+};
+
+class AttrList : public ClassList {
+protected:
+    AttrList(uint8_t *buff, size_t buff_len, bool parse) : ClassList(buff, buff_len, parse) {}
+
+public:
+    virtual ~AttrList()                  = default;
+    virtual eWscMessageType type()       = 0;
+    virtual bool init(const config &cfg) = 0; // used for create
+    bool init();                              // used for parse
+
+    static std::shared_ptr<AttrList> create(ieee1905_1::tlvWsc &tlv, const config &cfg);
+    static std::shared_ptr<AttrList> parse(ieee1905_1::tlvWsc &tlv);
+
+    template <class T> std::shared_ptr<T> getAttr() { return this->getClass<T>(); };
+    template <class T> std::shared_ptr<T> addAttr() { return this->addClass<T>(); };
+    bool finalize() { return ClassList::finalize(); };
+    size_t len() { return this->getMessageLength(); };
+    uint8_t *buffer() { return this->getMessageBuff(); };
+
+    // getter functions for common attributes for m1 and m2
+    std::string manufacturer() { return getAttr<cWscAttrManufacturer>()->manufacturer_str(); };
+    std::string model_name() { return getAttr<cWscAttrModelName>()->model_str(); };
+    std::string device_name() { return getAttr<cWscAttrDeviceName>()->device_name_str(); };
+    std::string serial_number() { return getAttr<cWscAttrSerialNumber>()->serial_number_str(); };
+    uint16_t encr_type_flags()
+    {
+        return getAttr<cWscAttrEncryptionTypeFlags>()->encr_type_flags();
+    };
+    uint16_t auth_type_flags()
+    {
+        return getAttr<cWscAttrAuthenticationTypeFlags>()->auth_type_flags();
+    };
+    uint8_t *enrollee_nonce() { return getAttr<cWscAttrEnrolleeNonce>()->nonce(); };
+    uint8_t *public_key() { return getAttr<cWscAttrPublicKey>()->public_key(); };
+    uint16_t rf_bands() { return getAttr<cWscAttrRfBands>()->bands(); };
+    bool valid();
+
+protected:
+    sWscAttrHeader *getNextAttrHdr()
+    {
+        return reinterpret_cast<sWscAttrHeader *>(
+            m_class_vector.empty() ? getMessageBuff() : m_class_vector.back()->getBuffPtr());
+    }
+    uint16_t getNextAttrType() { return ntohs(getNextAttrHdr()->type); };
+    size_t getRemainingBytes()
+    {
+        return m_class_vector.empty() ? getMessageBuffLength()
+                                      : m_class_vector.back()->getBuffRemainingBytes();
+    };
+    virtual bool valid_custom() = 0;
+
+private:
+    static eWscMessageType wsc_msg_type_from_buffer(uint8_t *buff, size_t buff_len);
```

Comment:
Any reason to not convert this method to static function in the cpp file?
I prefer using stand-alone static functions as it cleans up the header file (which is the actual interface to the class).

Reply:
Also, the input pointer type should probably be `void*`. Can prevent casting issues.